tools/java/org/unicode/cldr/api/CldrDataType.java

```diff
@@ -0,0 +1,204 @@
+package org.unicode.cldr.api;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import org.unicode.cldr.util.DtdData;
+import org.unicode.cldr.util.DtdData.Attribute;
+import org.unicode.cldr.util.DtdData.Element;
+import org.unicode.cldr.util.DtdType;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.function.Function.identity;
+import static org.unicode.cldr.util.DtdData.AttributeStatus.distinguished;
+import static org.unicode.cldr.util.DtdData.AttributeStatus.value;
+import static org.unicode.cldr.util.DtdData.Mode.OPTIONAL;
+
+/**
+ * Data types for non-locale based CLDR data. For the canonical specification for LDML data can
+ * be found at <a href="https://unicode.org/reports/tr35">Unicode Locale Data Markup Language<\a>.
+ */
+public enum CldrDataType {
+    /**
+     * Non-locale based BCP47 data, typically associated with international identifiers such as
+     * currency symbols, timezone identifiers etc.
+     */
+    BCP47(DtdType.ldmlBCP47, false, false),
+    /**
+     * Non-locale based supplemental data, typically associated with character tables (e.g. for
+     * break iterator).
+     */
+    SUPPLEMENTAL(DtdType.supplementalData, false, false),
+    /**
+     * Locale based LDML data consisting of internationalization information and translations on a
+     * per locale basis. LDML data for one locale may be inherited from other locales.
+     */
+    LDML(DtdType.ldml, true, true, DtdType.ldmlICU);
```

Comment:
why is ICU needed here?

Reply:
This was added recently since I have to support the "specials" directory, and I have to know things like "which elements are leaf elements" or "which attributes are valid". I'd rather not add it, but it allowed me to remove a whole pile of hard coded hacks if I just had access to the DTD.

The simple fact is that CLDR uses semantics above and beyond just XML, and things like CldrPath need that additional knowledge to work cleanly. I'd happily remove this if we can figure out a way to not need it, but for now it's definitely the best way to keep the API behaviour clean.

Reply:
From an API perspective, we shouldn't expect the API to enforce semantics that aren't representable in the DTD for specials.  So if someone wants to use the ShoeSize.dtd extension that you've never seen before, any semantics beyond what the DTD can enforce shouldn't be enforced by this API. 

Can't the DTD be read from the actual XML?

Reply:
That's sadly impossible at the moment.

The DTD can be read, but the DTD extensions are CLDR specific and the code that parses any special DTD extensions is hidden away in DtdData, which is in turn hooked to the DtdType enum, which cannot be arbitrarily extended.

It could all be done, but would require significant rewriting of underlying classes (rewrite DtdData, remove the DtdType enum in favour of some extensible ID class, hook DTD parsing into the core XML parser, probably rewrite chunks of CLDRFile etc.).

However this is all completely out-of-scope for my work, so I don't intend to consider trying to attempt it. If you feel strongly that it should be done, convince Mark to schedule another followup project for it.

Reply:
Could we just perform no validation on specials…?  what do you think, @macchiati ? If this is really an API, it shouldn't be just a shim to enable ICU.

Reply:
That's what my code used to do, but it doesn't work when you try and make "specials" work.

I need to be able to identify distinguishing attributes on _all_ path elements, since the CldrPath instance _only_ keeps those, and must be unique per visitation.

If you assume unknown attributes are "value" attributes, you get the same "special" path multiple times (because what should be a distinguishing attribute isn't present) but if you assume unknown attributes are distinguishing, you allow for duplicates to be treated differently (which messes up inheritance).

The semantics buried at the heart of CLDR are more than can be expressed in "normal" XML, and (without a lot of rewriting) you need the core libraries to understand all the DTDs that might be encountered. It's not a code problem so much as a fundamental data semantics problem.

Reply:
Also when you say ""If this is really an API, it shouldn't be just a shim to enable ICU"", that's not true, since the entire remit of this work _is_ to make a shim that mostly just supports ICU.

There's nothing stopping people reworking all this once it's in and once people have decided what they want the semantics for these extensions to be, but right now, my stated goal is _only_ to get ICU and a small number of Google internal special cases to work.